### PR TITLE
DYN-4880: Dynamo crash upon clicking Packages > Arkance Systems Settings

### PR DESCRIPTION
**Pull request by sm6srw**
### Purpose

We can't prevent the crash in DYN-4880 but we can improve our messaging when users install packages that targets other hosts.

image

Normal case: Warn if a package target other hosts but not the current host
Special case: Sandbox, warn if a package target only one known host (a conservative way to not warn for packages that works for all hosts). 

### Declarations

Check these if you believe they are true

- [X] The codebase is in a better state after this PR
- [X] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [X] The level of testing this PR includes is appropriate
- [X] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [X] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated

### Release Notes

Improves messaging when a user install a package that target a different environment.

### Reviewers

(FILL ME IN) Reviewer 1  (If possible, assign the Reviewer for the PR)

(FILL ME IN, optional) Any additional notes to reviewers or testers.

### FYIs

(FILL ME IN, Optional) Names of anyone else you wish to be notified of

### CODE REVIEW

**Review by mjkkirschner — APPROVED** [review]

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs [hunk 1]
```diff
@@ -5,6 +5,7 @@
 using System.IO;
 using System.Linq;
 using System.Reflection;
+using System.Runtime.CompilerServices;
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs:8]: is this needed?

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs [hunk 2]
```diff
@@ -804,6 +809,42 @@ internal async void ExecutePackageDownload(string name, PackageVersion package,
...
 
+                // determinate if any of the packages are targeting other hosts
+                var containsPackagesThatTargetOtherHosts = false;
+
+
+                // Known hosts
+                var knownHosts = Model.GetKnownHosts();
+
+                // Sandbox, special case: Warn if any package targets only one known host
+                if (String.IsNullOrEmpty(Host))
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs:820]: yeah, it's too bad the "Hosts" entry is not super well defined, and that core or sandbox does not exist. I think this logic will probably need to be flexible.

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs (hunk 1 above)
Comment (sm6srw) [reply src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs:8]: Thanks, nop, will remove.

File: src/DynamoPackages/PackageManagerClient.cs [hunk 3]
```diff
@@ -135,7 +135,7 @@ internal virtual PackageVersion GetPackageVersionHeader(string id, string versio
         /// Make a call to Package Manager to get the known
         /// supported hosts for package publishing and filtering
         /// </summary>
-        internal IEnumerable<string> GetKnownHosts()
+        internal virtual IEnumerable<string> GetKnownHosts()
```
Comment (pinzart90) [inline src/DynamoPackages/PackageManagerClient.cs:138]: why make it virtual ?

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs (hunk 2 above)
Comment (sm6srw) [reply src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs:818]: Yep, a have tried to be conservative and avoid false positives in sandbox. Our generic packages tends to be marked as targeting no known hosts or all known hosts. A package is most likely a "works on this host only" package if it targets only one host.

File: src/DynamoPackages/PackageManagerClient.cs (hunk 3 above)
Comment (sm6srw) [reply src/DynamoPackages/PackageManagerClient.cs:138]: So I can mock it.

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs [hunk 4]
```diff
@@ -804,6 +808,41 @@ internal async void ExecutePackageDownload(string name, PackageVersion package,
...
+                var containsPackagesThatTargetOtherHosts = false;
+
+                // Known hosts
+                var knownHosts = Model.GetKnownHosts();
+
+                // Sandbox, special case: Warn if any package targets only one known host
+                if (String.IsNullOrEmpty(Host))
+                {
+                    containsPackagesThatTargetOtherHosts =
+                        newPackageHeaders.Any(y => y.host_dependencies != null && y.host_dependencies.Intersect(knownHosts).Count() == 1);
```
[the remark this answers is not in the document]
Comment (sm6srw) [reply src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs:821]: I want this check to also happen to the initial package so I think this is OK.

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs [hunk 5]
```diff
@@ -804,6 +808,41 @@ internal async void ExecutePackageDownload(string name, PackageVersion package,
...
+                }
+                else
+                {
+                    // Warn if there are packages targeting other hosts but not our host
+                    var otherHosts = knownHosts.Except(new List<string>() {Host});
+                    containsPackagesThatTargetOtherHosts = newPackageHeaders.Any(x =>
+                    {
+                        // Is our host in the list?
+                        // If not, is any other host in the list?
+                        return x.host_dependencies != null && !x.host_dependencies.Contains(Host) && otherHosts.Any(y => x.host_dependencies != null && x.host_dependencies.Contains(y));
```
Comment (aparajit-pratap) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs:831]: aren't you checking for this condition twice in the same line: `x.host_dependencies != null`?

**Pull request merged**